Modules/_asynciomodule.c

```diff
@@ -429,19 +425,10 @@ future_schedule_callbacks(asyncio_state *state, FutureObj *fut)
            callbacks from the 'fut_callbacks' list. */
     }
 
-    if (fut->fut_callbacks == NULL) {
-        /* No more callbacks, return. */
-        return 0;
-    }
-
-    len = PyList_GET_SIZE(fut->fut_callbacks);
-    if (len == 0) {
-        /* The list of callbacks was empty; clear it and return. */
-        Py_CLEAR(fut->fut_callbacks);
-        return 0;
-    }
-
-    for (i = 0; i < len; i++) {
+    // Beware: An evil call_soon could change fut->fut_callbacks.
+    for (Py_ssize_t i = 0;
+         fut->fut_callbacks != NULL && i < PyList_GET_SIZE(fut->fut_callbacks);
```

Comment:
Pire python version works with the callbacks list copy:
```
        callbacks = self._callbacks[:]
        if not callbacks:
            return

        self._callbacks[:] = []
        for callback, ctx in callbacks:
            self._loop.call_soon(callback, self, context=ctx)
```
I think that C version should do the same:
1. Assign `fut->fut_callbacks` to temporary variable, set `fut->fut_callbacks` to `NULL`
2. Iterate over temporary saved list, `call_soon` every element
3. Decref temporary variable.